### DISCUSSION TITLE
Issue w/c_lib on primos v22

### DISCUSSION THREAD

**Discussion by WA6YDQ** [prirun/p50em#33] [General]
Hey all - 

While running a benchmark on the p50 emulator running Primos v22 I came across an error I didn't expect to see. Maybe I'm doing something wrong, and am looking for guidance.

I won't bore you with the benchmark program, but what I found was the exp() function returns 0 for negative values. Here is example code and results:

/* t2.c - test exp() function
k theis 10 2021
compile w/ ci t2 -ansi
filmem
link w/ bind -li ccmain -lo t2 -li c_lib -li
*/

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

double val, res;

int main(int argc, char *argv[]) {
int n=0;
    val = -2.0;
    for (n=0; n<10;n++) {
        res = exp(val);
        printf("exp(%10.4f) = %10.4f\n",val,res);
        val += 0.5;
    }

    return 0;
}

When run on the emulator I get these results:
OK, r t2
exp(   -2.0000) =     0.0000
exp(   -1.5000) =     0.0000
exp(   -1.0000) =     0.0000
exp(   -0.5000) =     0.0000
exp(    0.0000) =     1.0000
exp(    0.5000) =     1.6487
exp(    1.0000) =     2.7183
exp(    1.5000) =     4.4817
exp(    2.0000) =     7.3891
exp(    2.5000) =    12.1825
OK,

When run on my linux box with glibc I get:
 $ ./t2
exp(   -2.0000) =     0.1353
exp(   -1.5000) =     0.2231
exp(   -1.0000) =     0.3679
exp(   -0.5000) =     0.6065
exp(    0.0000) =     1.0000
exp(    0.5000) =     1.6487
exp(    1.0000) =     2.7183
exp(    1.5000) =     4.4817
exp(    2.0000) =     7.3891
exp(    2.5000) =    12.1825

My calculator and other devices give me the 2nd results as well. 

Am I doing something wrong compiling or linking? Am I missing something?

BTW: the whetstone benchmark shows my emulation runs about 16.5 MIPS.

- Kurt
-

**Comment by drboone** [comment]
The C math library stuff just calls the FORTRAN library stuff.
```
OK, slist ponent.ftn
      INTEGER N
      REAL*8 VAL
      REAL*8 RES

      N = 0
      VAL = -2.0

      DO 10 I=0,9
            RES = EXP(VAL)
            WRITE (1,20) VAL, RES
            VAL = VAL + 0.5
10    CONTINUE

20    FORMAT (4Hval=, F10.4, 5H res=, F10.4)

      STOP
      END
OK, r ponent
val=   -2.0000 res=    0.0000
val=   -1.5000 res=    0.0000
val=   -1.0000 res=    0.0000
val=   -0.5000 res=    0.0000
val=    0.0000 res=    1.0000
val=    0.5000 res=    1.6487
val=    1.0000 res=    2.7183
val=    1.5000 res=    4.4817
val=    2.0000 res=    7.3891
val=    2.5000 res=   12.1825

**** STOP
```
So it appears the usual libraries don't support negative exponents.  Boo.  Hiss.

What platform gives you 16.5?

**Reply by WA6YDQ** [reply]
Appreciate you all looking into this. 

- What platform gives you 16.5?

The platform I have the emulator on is an HP proliant server (24 cores Intel XEON running 3.3 GHz, 32GB ram, running ubuntu server) that is currently lightly loaded. I ran the emulator both from a hard drive (10,000 rpm RAID 0/5 etc) and from a ram disk (/run/user/1000). I don't see any speed difference between the hard drive performance and the ram performance. I expected MUCH faster running from ram.

The whetstone source is: https://github.com/varigit/arm_benchmarks/blob/master/whetstone/whetstone.c
I made a slight change in the .c source to allow the exp() portion to not error out. The performance decrease should be unnoticeable.

Emulator Results:
OK, r whet 1000

Loops: 1000, Iterations: 1, Duration: 6 sec.
C Converted Double Precision Whetstones: 16.7 MIPS
OK, r whet 10000

Loops: 10000, Iterations: 1, Duration: 56 sec.
C Converted Double Precision Whetstones: 17.9 MIPS
OK, r whet 50000

Loops: 50000, Iterations: 1, Duration: 280 sec.
C Converted Double Precision Whetstones: 17.9 MIPS
OK,


Running natively on the server:
kurt@server1:~ $ ./whet 100000

Loops: 100000, Iterations: 1, Duration: 3 sec.
C Converted Double Precision Whetstones: 3333.3 MIPS

Do I think these MIPS numbers are accurate? No idea. I expected the native numbers on the server to be right. 
Roy Longbottom says the 9950 prime was 12 MIPS.  

I'm sure there are other benchmarks that will give different results. 

- kurt

**Comment by hashbackup** [comment]
You might try running this on the other Prime emulator to make sure it isn't an emulator problem.

The emulator handles floating point by converting from Prime to IEEE, doing the operation, then converting from IEEE to Prime.  Wouldn't surprise me if there is a bug somewhere that maybe the Prime hardware diagnostics didn't catch or that I overlooked.

**Comment by drboone** [comment]
Hmm.  From Jan's emulator:
```
OK, r ponent
val=   -2.0000 res=    0.1353
val=   -1.5000 res=    0.2231
val=   -1.0000 res=    0.3679
val=   -0.5000 res=    0.6065
val=    0.0000 res=    1.0000
val=    0.5000 res=    1.6487
val=    1.0000 res=    2.7183
val=    1.5000 res=    4.4817
val=    2.0000 res=    7.3891
val=    2.5000 res=   12.1825

**** STOP
```

**Comment by hashbackup** [comment]
Yeah, as much scientific work was done on Primes, it would surprise me if their exp() library function was that broken.  It's nice having another reference to test these things without having to fire up a real Prime.  Thanks @janjaeger!
